wunderio/note
# Basic CRUD functionality

**Prior to running testing scenarios**
Ensure that Symfony environment is operational, whether run via Docker or not
If the environment is being run via Docker on your local machine, then ensure that all containers are booted, and then consider `localhost:8001` to be the domain/host name for the CRUD service

You can enter `localhost:8001/notes` and see if a grid of users appears to initially test this is true
If you see no data in this table, you need to add some data, via the the PUT endpoint (`PUT /notes/add`).
Alternatively, log into the `php` container in docker, and run `bin/console doctrine:fixtures:load` to get 10 randomly generated entries in the dev environment database (accessible env for `localhost:8001`)

Manual testing scenarios:

1. Create a note
Setup:
- Open HTTP client 'Postman'
- Create a new request with type `PUT`
- Enter domain `localhost:8001/notes/add`
- Open the 'query params' tab

Failure cases: 
 - Omit 'title' parameter to test that ParameterNotFoundException occurs when parameter 'title' is missing. Hit send
 - Enter no parameters, same exception. Hit send
 - Enter a text value that exceeds 255 characters. Hit send

Success case: 
 - Enter 'title' parameter, (key: 'title', value: [some_title]). Hit send
    - Result: JSON object returned with action completed and final data saved. Response code 201
 - Enter 'title' and 'text' parameter
    - Result: JSON object returned with action completed and final data saved. Response code 201

2. Update a note
Setup:
- Open HTTP client 'Postman'
- Create a new request with type `PATCH`
- Enter domain `localhost:8001/notes/:id`
- Open the 'query params' tab

Failure cases: 
- Enter no query parameters. Hit send.
    -  Result: ParameterNotFoundException
- Enter an id number (in path variables 'id') that does not exist in the note table
    - Result: EntityNotFoundException
Success cases:
- Enter a valid entity id in path variable 'id'. Enter a title and corresponding text value in query params
    - Result: JSON object returned with action completed and final data saved. Response code 200

3. Delete an entity
Setup:
- Open HTTP client 'Postman'
- Create a new request with type `DELETE`
- Enter domain `localhost:8001/notes/:id`

Failure case:
- Enter an id number (in path variables 'id') that does not exist in the note table
    - Result: EntityNotFoundException
- Enter a valid id number in path variable 'id'. Hit send
    - Result:  JSON object returned with action completed and final data saved. Response code 200

4. Get an entity
Setup:
- Open HTTP client 'Postman'
- Create a new request with type `GET`
- Enter domain `localhost:8001/notes/:id`

Failure case:
- Enter an id number (in path variables 'id') that does not exist in the note table
    - Result: EntityNotFoundException
- Enter a valid id number in path variable 'id'. Hit send
    - Result:  JSON object returned with action completed (get note) and final data saved. Response code 200

5. View all notes in table:
 - Open a modern Browser
 - Enter the URL `localhost:8001/notes`
 - Ensure page loads completely, and the table presented has items within it
 - Right click the page and press 'inspect', then select the console tab, ensuring there were no JavaScript errors on page load

Functionalities to test:
 - Pagination:
    - User can select amount of items to see in one pane at a given time
    - User can use arrow keys to navigate through the pages
 - Text search
    - User can utilise presented search bar to keyword search all items present in the dataset for specific string values. Only the 'text' field of the grid should be searched
 - Sort order
     - On page load, items presented should be presented sorted as per created_at field in descending order
     - User can click columns to change sorting order to depend on that column, and again to reverse sort direction
